sdk/identity/Azure.Identity/src/VisualStudioCodeCredential.cs

```diff
@@ -0,0 +1,146 @@
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using Microsoft.Identity.Client;
+
+namespace Azure.Identity
+{
+    /// <summary>
+    /// Enables authentication to Azure Active Directory using data from Visual Studio Code
+    /// </summary>
+    internal class VisualStudioCodeCredential : TokenCredential
+    {
+        private const string CredentialsSection = "VS Code Azure";
+        private const string ClientId = "aebc6443-996d-45c2-90f0-388ff96faa56";
+        private readonly IVisualStudioCodeAdapter _vscAdapter;
+        private readonly IFileSystemService _fileSystem;
+        private readonly CredentialPipeline _pipeline;
+        private readonly string _tenantId;
+
+        /// <summary>
+        /// Protected constructor for mocking
+        /// </summary>
+        protected VisualStudioCodeCredential() : this(default, default) {}
+
+        /// <inheritdoc />
+        public VisualStudioCodeCredential(string tenantId, TokenCredentialOptions options) : this(tenantId, options, default, default) {}
+
+        internal VisualStudioCodeCredential(string tenantId, TokenCredentialOptions options, IFileSystemService fileSystem, IVisualStudioCodeAdapter vscAdapter)
+        {
+            _tenantId = tenantId ?? "common";
+            _pipeline = CredentialPipeline.GetInstance(options);
+            _fileSystem = fileSystem ?? FileSystemService.Default;
+            _vscAdapter = vscAdapter ?? GetVscAdapter();
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
+            => await GetTokenImplAsync(requestContext, true, cancellationToken).ConfigureAwait(false);
+
+        /// <inheritdoc />
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+            => GetTokenImplAsync(requestContext, false, cancellationToken).EnsureCompleted();
+
+        private async ValueTask<AccessToken> GetTokenImplAsync(TokenRequestContext requestContext, bool async, CancellationToken cancellationToken)
+        {
+            using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("VisualStudioCodeCredential.GetToken", requestContext);
+
+            try
+            {
+                GetUserSettings(out var tenant, out var environmentName);
+
+                var cloudInstance = GetAzureCloudInstance(environmentName);
+                var storedCredentials = _vscAdapter.GetCredentials(CredentialsSection, environmentName);
+
+                try
+                {
+                    JsonDocument.Parse(storedCredentials);
+                    throw new CredentialUnavailableException("Need to re-authenticate user in VSCode Azure Account.");
+                }
+                catch (JsonException)
+                {
+                    var publicClient = (IByRefreshToken)PublicClientApplicationBuilder.Create(ClientId)
```

Comment:
You should use the MsalPublicClient abstraction rather than PublicClientApplication directly. This will help with mocking. If it doesn't meet your needs here we should extend it.

Reply:
If it's required that's ok for now. I'm refactoring how we use MSAL clients in our implementation based off some performance testing I'm doing, and this will be impacted by this in our next release anyway. However, I think you can accomplish this without an instance per call. The `AcquireTokenByRefreshTokenArgumentBuilder` probably have WithTenant and WithAuthorityh setters on it which will allow you set these per call.

Reply:
Fixed